components/camera/videosource/webcam.go

```diff
@@ -224,259 +134,253 @@ func NewWebcam(
+							c.buffer.release = nil
 						}
+
+						c.logger.Infow("camera reconnected")
+						c.mu.Unlock()
+						break reconnectLoop
 					}
 				}
 			}
 		}
 	})
 }
 
-func (c *webcam) Images(
-	ctx context.Context,
-	filterSourceNames []string,
-	extra map[string]interface{},
-) ([]camera.NamedImage, resource.ResponseMetadata, error) {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-	if err := c.ensureActive(); err != nil {
-		return nil, resource.ResponseMetadata{}, err
-	}
+// startBufferWorker starts a worker that continuously reads frames from the camera and writes them to the buffer.
+// When disconnected, it skips reading but continues running to resume immediately upon reconnection.
+func (c *webcam) startBufferWorker() {
+	c.mu.Lock()
+	frameRate := c.conf.FrameRate
+	c.mu.Unlock()
 
-	img, err := c.getLatestFrame()
-	if err != nil {
-		return nil, resource.ResponseMetadata{}, errors.Wrap(err, "monitoredWebcam: call to get Images failed")
-	}
+	interFrameDuration := time.Duration(float32(time.Second) / frameRate)
 
-	namedImg, err := camera.NamedImageFromImage(img, c.Name().Name, utils.MimeTypeJPEG, data.Annotations{})
-	if err != nil {
-		return nil, resource.ResponseMetadata{}, err
-	}
-	return []camera.NamedImage{namedImg}, resource.ResponseMetadata{CapturedAt: time.Now()}, nil
-}
+	c.workers.Add(func(ctx context.Context) {
+		ticker := time.NewTicker(interFrameDuration)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				c.logger.Debug("buffer worker context done")
+				return
+			case <-ticker.C:
+				c.mu.Lock()
 
-// ensureActive is a helper that guards logic that requires the camera to be actively connected.
-func (c *webcam) ensureActive() error {
-	if c.closed {
-		return errClosed
-	}
-	if c.disconnected {
-		return errDisconnected
-	}
-	return nil
-}
+				if c.disconnected {
+					c.mu.Unlock()
+					continue
+				}
 
-func (c *webcam) Image(ctx context.Context, mimeType string, extra map[string]interface{}) ([]byte, camera.ImageMetadata, error) {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
+				reader := c.reader
+				if reader == nil {
+					c.mu.Unlock()
+					continue
+				}
 
-	if err := c.ensureActive(); err != nil {
-		return nil, camera.ImageMetadata{}, err
-	}
-	if c.reader == nil {
-		return nil, camera.ImageMetadata{}, errors.New("underlying reader is nil")
-	}
-	img, err := c.getLatestFrame()
-	if err != nil {
-		return nil, camera.ImageMetadata{}, err
-	}
+				// Get the release function to call outside the lock to avoid potential deadlocks.
+				oldRelease := c.buffer.release
+				c.buffer.release = nil
+				c.mu.Unlock()
 
-	if mimeType == "" {
-		mimeType = utils.MimeTypeJPEG
-	}
-	imgBytes, err := rimage.EncodeImage(ctx, img, mimeType)
-	if err != nil {
-		return nil, camera.ImageMetadata{}, err
-	}
-	return imgBytes, camera.ImageMetadata{MimeType: mimeType}, nil
-}
+				// Call release and read outside the lock to avoid holding the lock during I/O
+				if oldRelease != nil {
+					oldRelease()
```

Comment:
Just want to make sure this strategy of releasing the last from just before reading the new frame works well

Would we not want to release immediately after retrieving the frame storing into the webcamBuffer?

Reply:
`Would we not want to release immediately after retrieving the frame storing into the webcamBuffer?` 

Like during an `Image`/`Images` call?

Reply:
I was asking about in this loop after we stuff the image into the buffer `c.buffer.frame = img`

But realizing we do not copy so release will wipe the frame, so we need the prev release right before capturing again.